tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/fusion_utils.h

```diff
@@ -0,0 +1,239 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_COMPILER_MLIR_HLO_INCLUDE_MLIR_HLO_DIALECT_MHLO_TRANSFORMS_FUSION_UTILS_H_
+#define TENSORFLOW_COMPILER_MLIR_HLO_INCLUDE_MLIR_HLO_DIALECT_MHLO_TRANSFORMS_FUSION_UTILS_H_
+
+#include <memory>
+#include <vector>
+
+#include "llvm/ADT/EquivalenceClasses.h"
+#include "llvm/Support/Debug.h"
+#include "mlir-hlo/Dialect/mhlo/IR/lhlo_ops.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"  // TF:llvm-project
+
+// This file implements some helper functions and classes used to do fusion
+// & code generation.
+
+namespace mlir {
+namespace lmhlo {
+
+// kLoop fusion template satisfies:
+//   - all ops in the fusion pattern are element-wise.
+//   - all the shapes of outputs of fusion pattern are same or have same number
+//   of elements, and thus can fit into a same parallel loop.
+//
+// kInput fusion template satisfies:
+//   - any op in the fusion pattern is either element-wise or a reduction.
+//   - if a op is a reduction, its output cannot be consumered by other
+//     ops in the same fusion pattern.
+//   - all the effective shapes of outputs of fusion pattern are same.
+//     - For element-wise op, its effective shape is its output shape.
+//     - For reduction op, its effective shape is its operand shape.
+enum FusionType {
+  // Not a fusion pattern
+  kNone,
+  // kLoop fusion pattern
+  kLoop,
+  // kInput fusion pattern and all reduce ops of the fused pattern are row
+  // reduction
+  kRowReduction,
+  // kInput fusion pattern and all reduce ops of the fused pattern are column
+  // reduction
+  kColReduction,
+};
+
+// Returns true if the op is an elementwise unary lmhlo op.
+// TODO: use fusibility interface
+bool isElementWiseUnary(Operation* op);
+
+// Returns true if the op is an elementwise binary lmhlo op.
+// TODO: use fusibility interface
+bool isElementWiseBinary(Operation* op);
+
+// Returns true if the op is an elementwise lmhlo op.
+// TODO: use fusibility interface
+bool isElementWise(Operation* op);
+
+// Returns true if this op is a rank-2 row reduction.
+bool isRank2RowReduction(Operation* op);
+
+// Returns true if this op is a rank-2 column reduction.
+bool isRank2ColReduction(Operation* op);
+
+// Returns true if the op may be fused with other ops.
+bool isFusible(Operation* op);
+
+// Returns the number of operands that are supposed to be written.
+// For some ops (e.g. lmhlo ops), some operands are the output memrefs
+// Thus these operands are supposed to be updated.
+int getNumResultOperands(Operation* op);
```

Comment:
Thanks! We need this since different lmhlo ops may have different number of output operands. This is a helper function to query the number of output operands. I'll change the implementation of this function to use MemoryEffect interface.